pennylane/templates/subroutines/select.py

```diff
@@ -0,0 +1,91 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the SELECT template.
+"""
+# pylint: disable=too-many-arguments
+
+import itertools
+import pennylane as qml
+from pennylane.operation import Operation
+from pennylane import math
+
+
+class SELECT(Operation):
+    r"""
+    Applies specific input operations depending on the state of
+    the designated control qubits.
+
+    .. math:: SELECT|X\rangle \otimes |\psi\rangle = |X\rangle \otimes U_x |\psi\rangle
+
+    Args:
+        ops (list[Operator]): operations to apply
+        control_wires (Sequence[int]): the wires controlling which operation is applied
+        id (str or None): String representing the operation (optional)
+
+    .. note::
+        The position of the operation in the list determines which qubit state implements that operation.
+        For example, when the qubit register is in the state :math:`|00\rangle`, we will apply ``ops[0]``. When the qubit register is
+        in the state :math:`|10>`, we will apply ``ops[2]``.
+
+    **Example**
+
+    >>> dev = qml.device('default.qubit',wires=4)
+    >>> ops = [qml.PauliX(wires=2),qml.PauliX(wires=3),qml.PauliY(wires=2),qml.SWAP([2,3])]
+    >>> @qml.qnode(dev)
+    >>> def circuit():
+    >>>     qml.SELECT(ops,control_wires=[0,1])
+    >>>     return qml.state()
+    ...
+    >>> print(qml.draw(circuit,expansion_strategy='device')())
+    0: ─╭○─╭○─╭●─╭●────┤  State
+    1: ─├○─├●─├○─├●────┤  State
+    2: ─╰X─│──╰Y─├SWAP─┤  State
+    3: ────╰X────╰SWAP─┤  State
+
+    """
+
+    num_wires = qml.operation.AnyWires
+
+    def __init__(self, ops, control_wires, id=None):
+        control_wires = qml.wires.Wires(control_wires)
+        if 2 ** len(control_wires) < len(ops):
+            raise ValueError(
+                f"Not enough control wires ({len(control_wires)}) for the desired number of "
+                + f"operations ({len(ops)}). At least {int(math.ceil(math.log2(len(ops))))} control "
+                + "wires required."
+            )
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control_wires
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+        for op in ops:
+            qml.QueuingManager.remove(op)
+
+        all_wires = sum(op.wires for op in ops) + control_wires
+        super().__init__(ops, control_wires, wires=all_wires, id=id)
+
+    @staticmethod
+    def compute_decomposition(
+        ops, control_wires, wires=None
+    ):  # pylint: disable=arguments-differ, unused-argument
+        states = list(itertools.product([0, 1], repeat=len(control_wires)))
```

Comment:
Afaik, `itertools.product` does give the output in the same order. An alternative approach could be this ugly
```
[tuple(map(int, list(bin(i)[2:].zfill(len(control_wires)))) for i in range(2**len(control_wires))]
```

Reply:
So something like this?
`[[int(i) for i in list(bitstring)] for bitstring in [format(i,f'0{len(control_wires)}b') for i in range(2**len(control_wires))]]`

Reply:
My concern is that if `itertools.product` return a different order, for any reason such as version change etc., then the ops will be silently applied to the wrong states. If it is guaranteed that `itertools.product` will not change the order in the future, then it can be safely used.